keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,720 @@
+
+  - `KUBE-KUBELET-CANARY`, which is used by the `utiliptables.Monitor`
+    functionality to notice when the iptables rules have been flushed
+    and kubelet needs to recreate its rules.
+
+The reason that the `MARK` chains exist is because most of
+kube-proxy's service-processing logic occurs in subchains of the
+`OUTPUT` and `PREROUTING` chains of the `nat` table (because those are
+the only places you can call the `DNAT` target from, to redirect
+packets from a service IP to a pod IP), but neither masquerading nor
+dropping packets can occur at that point in IPTables processing:
+dropping packets can only occur from chains in the `filter` table,
+while masquerading can only occur from the `POSTROUTING` chain of the
+`nat` table (because the kernel can't know what IP to masquerade the
+packet to until after it has decided what interface it will send it
+out on, which necessarily can't happen until after it knows if it is
+going to DNAT it). To work around this, when Kubernetes wants to
+masquerade or drop a packet, it just marks it as needing to be
+masqueraded or dropped later, and then one of the later chains handles
+the actual masquerading/dropping.
+
+This approach was not _necessary_; kube-proxy could have just
+duplicated its matching logic between multiple IPTables chains, eg,
+having a rule saying "a packet sent to `192.168.0.11:80` should be
+redirected to `10.0.20.18:8080`" in one chain and a rule saying "a
+packet whose original pre-NAT destination was `192.168.0.11:80` should
+be masqueraded" in another.
+
+But using `KUBE-MARK-MASQ` allows kube-proxy to keep its logic more
+efficient, compact, and well-organized than it would be otherwise; it
+can just create a pair of adjacent rules saying "a packet sent to
+`192.168.0.11:80` should be redirected to `10.0.20.18:8080`, and
+should also be masqueraded".
+
+#### Eliminating `KUBE-MARK-DROP`
+
+In theory, `KUBE-MARK-DROP` is useful for the same reason as
+`KUBE-MARK-MASQ`, although in practice, it turns out to be
+unnecessary. Kube-proxy currently drops packets in two cases:
+
+  - When using the `LoadBalancerSourceRanges` feature on a service,
+    any packet to that service that comes from outside the valid
+    source IP ranges is dropped.
+
+  - When using `Local` external traffic policy, if a connection to a
+    service arrives on a node that has no local endpoints for that
+    service, it is dropped.
+
+In both cases, the use of `KUBE-MARK-DROP` from the `nat` table could
+easily be replaced with direct calls to `-j DROP` in the `filter`
+table.
+
+In the `LoadBalancerSourceRanges` case, we currently create a
+per-service `KUBE-FW` chain in the `nat` table that checks against
+each allowed source range and forwards traffic from those ranges to
+the `KUBE-SVC` or `KUBE-XLB` chain for the service. At the end
+of the chain, it calls `KUBE-MARK-DROP` on any unmatched packets.
+
+But there is no reason why the `FW` chain could not simply be moved to
```

Comment:
I wonder if there's any side-effect here with third-party things that might add additional drop rules or change the default policy, which a unilateral ALLOW would break.

Reply:
Or maybe `-j RETURN` rather than allow?